### PR TITLE
Add offline support via Service Worker

### PR DESCRIPTION
**Pull request by marcoscaceres**
Adds offline support and notifies users if an update becomes available. 

Preview at: 
https://marcoscaceres.github.io/tpac2016/

### CODE REVIEW

**Comment by marcoscaceres** [comment]
The toast needs a little UI love, but offline works. 

You may just have a PWA for TPAC 🎉  

Maybe @jakearchibald should double check sw_registration.js and and sw.js?

**Comment by marcoscaceres** [comment]
(and yes, not including the toast dynamically is kinda crap... but I'm lazy)

File: assets/js/async.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
+'use strict';
```
Comment (marcoscaceres) [inline assets/js/async.js:1]: this is: http://github.com/marcoscaceres/async/

**Comment by tripu** [comment]
I'll let @guibbs review this one, since he's the father of the creature.

**Comment by gbaudusseau** [comment]
😍😍😍 on it!

**Comment by gbaudusseau** [comment]
So when I test it locally (and run it on chrome Nexus 5), update the app, the toast doesn't appear. It updates the app automatically.

**Comment by marcoscaceres** [comment]
Just checking, are you updating the service worker? You need to make a change to the service worker (e.g., change the version number on the cache) for the change to show in the toast. Also, the toast only shows if the browser detects a service worker has changed (e.g., via a reload). 

So to be clear, as you update the files of the app, you need to bump the version string inside the service worker. Also, don't stress if you don't see the toast. Navigation also triggers an update behind the scenes. The main thing is that users will be able to use the app offline. 

Note also, the JSON file that contains the calendar data is not cached right now, as the schedule code is still broken. Let's get that fixed and then we can make it work offline.

File: index.html [hunk 2]
```diff
@@ -133,7 +134,11 @@
       </nav>
     </main>
     <div id="mask"></div>
-
+    <aside id="update-toast">
+      <p>App has been updated!
+      <button id="refresh">Refresh?</button>
+      <button id="dismiss">No thanks!</button></p>
+    </aside>
     <script src="http://localhost:9091"></script>
```
Comment (jakearchibald) [inline index.html:142]: Mistake?

File: sw.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    const cache = yield caches.open(SITE_CACHE);
+    yield cache.addAll(resources);
+  }));
+});
+
+self.addEventListener("activate", () => {
+  async.task(function*() {
+    const keys = yield caches.keys();
+    yield keys
+      .filter(key => key !== SITE_CACHE)
```
Comment (jakearchibald) [inline sw.js]: Is this site going to be the only cache-user on the origin? Just making sure other projects' caches aren't being deleted at this point.

File: sw.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    try {
+      const netResponse = yield fetch(ev.request);
+      if (netResponse.ok) {
+        return netResponse;
+      }
+    } catch (err) {
+      // just return the index if all goes bad.
+      console.error(err, ev.request.url);
+    }
+    return yield caches.match("/");
```
Comment (jakearchibald) [inline sw.js:98]: Returning the root page if a network request fails?

File: sw.js (hunk 3 above)
Comment (marcoscaceres) [reply sw.js]: Good catch. As there are no other ones, this should be ok.

File: index.html (hunk 2 above)
Comment (marcoscaceres) [reply index.html:142]: Site still under development by w3c.

File: sw.js (hunk 4 above)
Comment (marcoscaceres) [reply sw.js:98]: As the site won't be showing any external content, it should be ok for now. If there is time, we should add proper fallback for each type (e.g., 404.svg|png|etc.).

File: sw.js (hunk 3 above)
Comment (marcoscaceres) [reply sw.js]: I added a guard: 

``` JS
key.startWith("skeleton-")
```

**Comment by gbaudusseau** [comment]
Can I merge that one @marcoscaceres ? Or do you want to wait for me to fix the calendar script? Im currently working on it.

**Comment by marcoscaceres** [comment]
Up to you, but once you do, be mindful that the SW is going to be taking over, and the js will be served via the cache API. This means you will need to "shift-refresh" to see any changes to files (by bypassing the SW) - or you need to bump the version in the sw.js file and refresh. If that gets annoying, temporarily remove the broken js file from the sw.js file + clear the "cache storage" in Chrome's dev tools.

**Pull request closed without merging**